crates/core/transaction/src/is_action.rs

```diff
@@ -367,9 +367,40 @@ impl IsAction for Swap {
                         .cloned(),
                 }
             }
-            None => SwapView::Opaque {
-                swap: self.to_owned(),
-            },
+            None => {
+                // If we can find a matching BSOD in the TxP, use it to compute the output notes
+                // for the swap.
+                let bsod = txp
+                    .batch_swap_output_data
+                    .iter()
+                    // This finds the first matching one; there should only be one
+                    // per trading pair per block and we trust the TxP provider not to lie about it.
+                    .find(|bsod| bsod.trading_pair == self.body.trading_pair);
+
+                let denom_1 = txp.denoms.get(&self.body.trading_pair.asset_1()).cloned();
+                let denom_2 = txp.denoms.get(&self.body.trading_pair.asset_2()).cloned();
+
+                SwapView::Opaque {
+                    swap: self.to_owned(),
+                    batch_swap_output_data: bsod.cloned(),
+                    asset_1_metadata: denom_1.clone(),
+                    asset_2_metadata: denom_2.clone(),
+                    output_1: Some(
+                        Value {
+                            amount: self.body.delta_1_i,
```

Comment:
In the `SwapPlaintext` Protobuf type, `delta_1_i` and `delta_2_i` refer to the _inputs_ of the swap, not the outputs. So I _believe_ this is incorrect.

That said, I don't know how you would get outputs on an opaque swap. For a visible swap, you can call `swap_plaintext.output_notes(bsod)`, but `swap_plaintext` isn't available here. So I'm not sure how to calculate the outputs correctly 😬

Reply:
Yes, this seems incorrect. $\Delta_{1,i}$ and $\Delta_{2,i}$ are the inputs. But since the outputs are the pro-rata shares of the batch output, which is public, and the inputs are public, it's possible to compute the output amounts.

It's indeed not possible to compute the output _notes_, since they rely on secret information: a note is a value plus a control address plus a blinding factor, and the control address and blinding factors are private. That's why in the `SwapClaimView` the outputs are `NoteView`s, but here they're mere `ValueView`s. 

(Side note: is that actually true? It looks like this is constructing a `Value`, it should be a `ValueView`...)

The method for computing the output amounts (used internally by `output_notes`) is `BatchSwapOutputData::pro_rata_outputs`: https://rustdoc.penumbra.zone/main/penumbra_dex/struct.BatchSwapOutputData.html#method.pro_rata_outputs